File: src/C/b16c/b16c.c
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <isopt.h>

int main(int argc, char *argv[]) {
	int i;
	int DECODE = 0;
	for (i = 0; i < argc; i++) {
		if (!isopt(argv[i])) {
			if (strcmp(argv[i], "-h") == 0 | strcmp(argv[i], "--help") == 0) {
				system("yldoc b16c");
				return 0;
			} else if (strcmp(argv[i], "-v") == 0 | strcmp(argv[i], "--version") == 0) {
				printf("Version 1 in C\n");
				return 0;
			} else if (strcmp(argv[i], "-d") == 0 | strcmp(argv[i], "--decode") == 0) {
				DECODE = 1;
			}
		}
	}
	char a;
	if (DECODE) {
		char b;
		while (EOF != (a = getchar())) {
			b = getchar();
			putchar((a - 65) * 16 + b - 65);
		}
	} else {
		while (EOF != (a = getchar())) {
			putchar(a / 16 + 65);
			putchar(a - a / 16 * 16 + 65);
		}
	}
	return 0;
}

File: src/DEPRECATED/bin/exec/b16c.py
#!/usr/bin/env python
'''
b16c in Python. Should get same output as in C.
'''

import os
import struct
import sys

from linuxminipy.libisopt import isopt

VERSION = 2.3


# T#O#D#O: Still bugs in MSYS2

def main():
    decode = False
    for sysarg in sys.argv[1:]:
        if isopt(sysarg):
            if sysarg in ('-h', '--help'):
                os.system('yldoc b16c')
                sys.exit(0)
            elif sysarg in ('-v', '--version'):
                print(str(VERSION) + ' in Python')
                sys.exit(0)
            elif sysarg in ('-d', '--decode'):
                decode = True
    if decode:
        f = open(sys.stdin.fileno(), mode='rb')
        o = open(sys.stdout.fileno(), mode='wb')
        while True:
            inp1 = f.read(1)
            inp2 = f.read(1)
            if not inp1 or not inp2: break
            o.write(struct.pack('b', (ord(inp1) - 65) * 16 + ord(inp2) - 65 - 128))
    else:
        f = open(sys.stdin.fileno(), mode='rb')
        o = open(sys.stdout.fileno(), mode='w')
        while True:
            inp = f.read(1)
            if not inp: break
            inp = ord(inp)
            o.write(chr(inp // 16 + 65) + chr(inp % 16 + 65))
    o.close()
    f.close()


if __name__ == '__main__':
    main()
